runtime: run the QEMU VMM process with a non-root user

fengwang666 opened the pull request:
### Description ###
The change is to enable running QEMU VMM process as a non-root user in Kata for better security. A new flag `rootless` is added to the hypervisor config. If the flag is set to true, when the kata shim v2 is creating a Pod sandbox, it creates a random user (e.g. qemu-1180) and uses it to launch the QEMU process. It also creates a file directory structure (`/run/user/[uid]/vc/...`) and use it to store the vmm files (such as qmp.sock, vhost-fs.sock, console.sock). The file directory is owned by the qemu user. This change also utilizes the [existing rootless code path](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/pkg/rootless/rootless.go).

### Testing ###
Manually tested the change. I was able to manually build the kata-runtime and use it to launch pods and containers using `crictl`, following [the doc](https://github.com/kata-containers/documentation/blob/master/how-to/containerd-kata.md#launch-pods-with-crictl-command-line).  
Manually tested this change in an AWS EKS cluster. The limitations of this feature are documented in [another PR](https://github.com/kata-containers/kata-containers/pull/2546).
 
### Questions ###
1. Does this approach look acceptable? Any better suggestions? Any use case this will break?
2. In the current change I use a random run-as user for QEMU process based on the assumption that it shouldn't matter to the end customers which run-as is used as long as it's not root. I also attempted to use the uid/gid from the linux security context in the pod spec and it also works, but then the end user is put on the burden to specify the non-root uid/gid for QEMU. Any preference on this two options?
3. Any other changes I need to make before this can be accepted?

### Related Issues ###
https://github.com/kata-containers/kata-containers/issues/2444

bergwolf commented:
> it creates a random user (e.g. qemu-1180) and uses it to launch the QEMU process

How would you grant the random user access to devices like `/dev/kvm` and `/dev/vhost-xxx` ?

fengwang666 commented:
> How would you grant the random user access to devices like /dev/kvm and /dev/vhost-xxx ?

@bergwolf , for `/dev/kvm`, I added the group `kvm` to the QEMU process as a [supplemental group](https://github.com/fengwang666/kata-containers/blob/a20194b754df09854ef272a6c289084526205dec/src/runtime/containerd-shim-v2/create.go#L316-L323). 

I didn't do anything for `/dev/vhost-net` and `/dev/vhost-vsock`. It's tricky because by default both devices are owned by `root:root` and the permission is 600. How does QEMU process access these two device nodes? Or what test can I run to verify something is broken?

fengwang666 commented:
> > How would you grant the random user access to devices like /dev/kvm and /dev/vhost-xxx ?
> 
> @bergwolf , for `/dev/kvm`, I added the group `kvm` to the QEMU process as a [supplemental group](https://github.com/fengwang666/kata-containers/blob/a20194b754df09854ef272a6c289084526205dec/src/runtime/containerd-shim-v2/create.go#L316-L323).
> 
> I didn't do anything for `/dev/vhost-net` and `/dev/vhost-vsock`. It's tricky because by default both devices are owned by `root:root` and the permission is 600. How does QEMU process access these two device nodes? Or what test can I run to verify something is broken?

It looks like QEMU process doesn't need to access `/dev/vhost-vsock` device. Kata runtime already passes its [FD](https://github.com/fengwang666/kata-containers/blob/2859600a6fad3ccf6abb95144d2e9136b93a6e6f/src/runtime/virtcontainers/qemu_arch_base.go#L551) to the QEMU process. The QEMU command input also confirms that: `-device vhost-vsock-pci,disable-modern=false,vhostfd=3,id=vsock-962631188,guest-cid=962631188`. Am I miss anything?

fidencio commented:
@fengwang666, are you using virtiofsd?  Wouldn't that require a root QEMU?
I like the idea of your PR, but we need to consider every kata-containers usecase carefully.

fengwang666 commented:
@fidencio , I'm using the default setting during the test and it looks like it's using virtiofsd. When I run `ps` command, I can see two virtiofsd processes, which still run as root as expected. Any specific test case do you suggest?

bergwolf commented:
> > How would you grant the random user access to devices like /dev/kvm and /dev/vhost-xxx ?
> 
> @bergwolf , for `/dev/kvm`, I added the group `kvm` to the QEMU process as a [supplemental group](https://github.com/fengwang666/kata-containers/blob/a20194b754df09854ef272a6c289084526205dec/src/runtime/containerd-shim-v2/create.go#L316-L323).
> 
Ok, the suplemental group name should be configuration.

> I didn't do anything for `/dev/vhost-net` and `/dev/vhost-vsock`. It's tricky because by default both devices are owned by `root:root` and the permission is 600. How does QEMU process access these two device nodes? Or what test can I run to verify something is broken?

qemu needs to access /dev/vhost-net for networking and /dev/vhost-vsock for vsock communication to kata-agent in the guest.

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,80 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+    if err != nil {
+        return err
+    }
+    gid, err := strconv.Atoi(u.Gid)
+    if err != nil {
+        return err
+    }
+	runtimeConfig.HypervisorConfig.Uid = uint32(uid)
+	runtimeConfig.HypervisorConfig.Gid = uint32(gid)
```
you have some problems with the indentation

fengwang666 replied:
My bad. I forgot golang uses tab instead of space 😅

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,80 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+			return err
+		}
+    }
+    if dir != nil && !dir.IsDir() {
+    	return fmt.Errorf("%s is expected to be a directory", rootDir)
+	}
+
+	if err := os.Setenv("XDG_RUNTIME_DIR", rootDir); err != nil {
+		return err
+	}
```
same here

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,80 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+    maxAttempt := 5
+    attempt := 0
+
+    for ok := true; ok; ok = err != nil && attempt < maxAttempt {
+        userName = fmt.Sprintf("qemu-%v", rand.Intn(100000))
+        _, err = utils.RunCommand([]string{"useradd", userName})
+		attempt++
+        if err != nil {
```
please run `gofmt` on all the file

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,80 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
 }
+
+func configureNonRootHypervisor(runtimeConfig *oci.RuntimeConfig) error {
+	userName, err := createVmmUser()
```
I'm afraid, if something fails, you'll end up leaking a lot of users into the system, maybe you should add a `defer` to check the error

```go
defer func() {
    if err != nil {
        removeVmmUser()
    }
}()

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,80 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+
+    for ok := true; ok; ok = err != nil && attempt < maxAttempt {
+        userName = fmt.Sprintf("qemu-%v", rand.Intn(100000))
+        _, err = utils.RunCommand([]string{"useradd", userName})
```
I consider that this is prone to hacks if the PATH environment variable is not configured correctly, for security reasons I'd prefer to use the full path `/usr/sbin/useradd`

devimc replied:
this command may need more options, like `-M -N -s /bin/rbash` (and maybe create `/bin/rbash` if it doesn't exist)

devimc reviewed (commented):
thanks @fengwang666

fengwang666 commented:
@devimc I have addressed your comments. Can you also approve to run the workflows? Thanks!

devimc commented on src/runtime/vendor/github.com/kata-containers/govmm/qemu/qemu.go:
```diff
@@ -2439,6 +2439,13 @@ type Config struct {
 	// Ctx is the context used when launching qemu.
```
please raise a PR in https://github.com/kata-containers/govmm/ for this change

fengwang666 replied:
@devimc, thanks! PR raised: https://github.com/kata-containers/govmm/pull/197

egernst replied:
For the sake of this PR (review), can you make the vendor hack/addition a separate commit?

fengwang666 replied:
@egernst I just did as suggested. Can you please approve the workflow to run?

devimc commented:
> Can you also approve to run the workflows?

done

fengwang666 commented:
/test
/test-vfio

egernst commented on src/runtime/virtcontainers/utils/utils.go:
```diff
@@ -390,3 +391,63 @@ outer:
 	return nil
 }
+
+func MkdirAllWithInheritedOwner(path string, perm os.FileMode) error {
```
Can the utils additions be added as their own commit, and can we add unit tests?

fengwang666 replied:
I previously added the unit tests but removed them because it requires root to run them (for chown). According to https://github.com/kata-containers/kata-containers/issues/2446, root is removed in the github action.

jodh-intel replied:
@fengwang666 - understood, but that's a temporary bug. We do need tests for this function so you can add them and protect them using a simple approach like this:

- https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/mount_test.go#L307..L309

Or for more elaborate handling, you can use:

- https://github.com/kata-containers/kata-containers/tree/main/src/runtime/pkg/katatestutils

fidencio replied:
Hey, root tests have been re-enabled by a PR merged earlier Today.

egernst commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,99 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
 	return nil
 }
+
+func configureNonRootHypervisor(runtimeConfig *oci.RuntimeConfig) error {
```
I think there'll be more to discuss/review, but we'll want to make sure we have a unit test added for this function as well (ie, call function, verify what's put onto filesystem...

fengwang666 replied:
I can add an integration tests as a follow up. Create an [issue](https://github.com/kata-containers/kata-containers/issues/2607) to track it and please assign it to me.

egernst commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,99 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+	runtimeConfig.HypervisorConfig.Uid = uint32(uid)
+	runtimeConfig.HypervisorConfig.Gid = uint32(gid)
+
+	userTmpDir := path.Join("/run/user/", fmt.Sprint(uid))
```
I have some concerns that we are now leaving artifacts at yet another location on the filesystem.

Today, we leave artifacts for a running pod at:
/run/vc/
/run/kata-containers

As far as I understand, instead of leaving artifacts at `/run/vc/vm/`, we'll now leave them at `/run/user/<user id>/` when using this configuration?

In a perfect world, we'd have everything under `/run/kata-containers/` ...

Is the /run/user/<> part required in order to ensure appropriate permissions?

fengwang666 replied:
> Is the /run/user/<> part required in order to ensure appropriate permissions?

That's correct. Both `/run/vc` and `/runc/kata-containers` are only accessible by root (probably for good reasons).

egernst commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,99 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+
+	rbashPath := "/bin/rbash"
+	if _, err := os.Stat(rbashPath); os.IsNotExist(err) {
+		if _, err = utils.RunCommand([]string{"ln", "/bin/bash", rbashPath}); err != nil {
```
Correct me if I'm misinterpreting, but, I don't think the runtime should be creating this link on the host, right?

fengwang666 replied:
I added this to address @devimc 's comment on using `rbash` and create it if it doesn't exist. What's the concern of creating rbash?

amshinde replied:
Do you really need this? Can you not great user without providing the login shell?

devimc replied:
@fengwang666 can you use  `/sbin/nologin` ?

egernst reviewed (commented):
First, thanks for diving into this @fengwang666!  Some initial feedback.

Sorry if I missed this; as a runtime admin, how do I configure this so I can use it?

How does this work with cloud-hypervisor, or other VMM solutions?

fengwang666 commented:
> First, thanks for diving into this @fengwang666! Some initial feedback.
> 
> Sorry if I missed this; as a runtime admin, how do I configure this so I can use it?
> 
> How does this work with cloud-hypervisor, or other VMM solutions?

This behavior can be enabled through `enable_rootless_vmm` flag in the hypervisor config. So far this only applies to QEMU. I think it can be extended to cloud-hypervisor and other vmm as well.

egernst commented:
/test

fengwang666 commented:
@egernst @devimc , I just realized the jenkins ci is not testing the new code path because `defaultVmmRootless` is `false`. I temporarily updated the `defaultVmmRootless`flag to `true` for now. I assume I will need to add proper flags in the ci repo before merging it.

devimc commented:
/test-ubuntu

bergwolf commented:
/test

devimc commented:
a lot of `Permission denied...` errors, for example:

```
Error: CreateContainer failed: QMP command failed: Could not open '/dev/loop0': Permission denied: unknown
```
```
ctr: failed to create shim: QMP command failed: vfio 0000:00:03.0: failed to open /dev/vfio/3: Permission denied: unknown
```

@fengwang666  I guess this is expected, right?  can you document its limitations?

amshinde commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,99 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+
+	for ok := true; ok; ok = err != nil && attempt < maxAttempt {
+		userName = fmt.Sprintf("qemu-%v", rand.Intn(100000))
+		_, err = utils.RunCommand([]string{"/usr/sbin/useradd", "-M", "-s", rbashPath, userName})
```
Again, this path may differ depending on the distro. Maybe just call into `useradd` rather than specifying the path.

devimc replied:
IMO `useradd` is prone to hacks if `PATH` variable is not configured correctly

devimc replied:
@fengwang666 my suggestion here is to defined a list of `useradd` paths (`/usr/sbin/useradd`, `/sbin/useradd`, ....), iterate over that list and run the command if it exist, what do you think?

devimc replied:
@fengwang666  wdyt?

devimc commented:
@fengwang666 what changes do I need in my configuration file / system to get this PR working?

fengwang666 commented:
> @fengwang666 what changes do I need in my configuration file / system to get this PR working?

No configuration change is needed.

devimc commented:
@fengwang666 how do you run it?

fengwang666 commented:
> @fengwang666 how do you run it?

@devimc Here is how I did the test locally:

1. Install `kata` from `kata-static-2.2.0-alpha1-x86_64.tar.xz` and `crictl` following the kata documentations.
2. run `make && sudo -E PATH=$PATH make install` in the `kata-containers/src/runtime/` directory to build and install the kata-runtime binary with the change. In this case the newly built `kata-runtime` replaces the the one installed in the step 1. 
3. Then I use `sudo crictl runp -r kata nonroot-pod2-config.json` to start a kata pod. `nonroot-pod2-config.json` is a trivial config. 

I'm on Slack Kata Container Channel and I'm happy to chat with you over there if that's more productive.

devimc commented:
thanks @fengwang666 

> Install kata from kata-static-2.2.0-alpha1-x86_64.tar.xz and crictl following the kata documentations.

why not `ctr` ?

>  Then I use sudo crictl runp -r kata nonroot-pod2-config.json to start a kata pod. nonroot-pod2-config.json is a trivial config.

could you share the `nonroot-pod2-config.json` ?

> I'm on Slack Kata Container Channel and I'm happy to chat with you over there if that's more productive.

I don't use slack.

fengwang666 commented:
> thanks @fengwang666
> 
> > Install kata from kata-static-2.2.0-alpha1-x86_64.tar.xz and crictl following the kata documentations.
> 
> why not `ctr` ?
> 
> > Then I use sudo crictl runp -r kata nonroot-pod2-config.json to start a kata pod. nonroot-pod2-config.json is a trivial config.
> 
> could you share the `nonroot-pod2-config.json` ?
> 
> > I'm on Slack Kata Container Channel and I'm happy to chat with you over there if that's more productive.
> 
> I don't use slack.

@devimc `ctr` should work too. I switched from `ctr` to `crictl` when I was trying to use the security context from the pod spec to set up the uid/gid and stuck with it. 

nonroot-pod2-config.json:
```
{
    "metadata": {
        "name": "nginx2-sandbox",
        "namespace": "default",
        "attempt": 1,
        "uid": "hdishd83djaidwnduwk28bcs3"
    },
    "log_directory": "/tmp",
    "linux": {
    }
}
```

devimc commented:
@fengwang666 I'm unable to get this PR working on my system

```
FATA[0000] run pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: failed to launch qemu: exit status 1, error messages from qemu log: Could not access KVM kernel module: Permission denied
qemu-system-x86_64: failed to initialize kvm: Permission denied
: unknown 
```

```
$ ls /dev/kvm -l
crw------- 1 root root 10, 232 Aug 18 16:22 /dev/kvm
````

I think is is something that we should document, btw https://github.com/kata-containers/documentation/pull/565/files is a PR from @jodh-intel  that never landed, you may want to used as base for your document

fengwang666 commented:
> @fengwang666 I'm unable to get this PR working on my system
> 
> ```
> FATA[0000] run pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: failed to launch qemu: exit status 1, error messages from qemu log: Could not access KVM kernel module: Permission denied
> qemu-system-x86_64: failed to initialize kvm: Permission denied
> : unknown 
> ```
> 
> ```
> $ ls /dev/kvm -l
> crw------- 1 root root 10, 232 Aug 18 16:22 /dev/kvm
> ```
> 
> I think is is something that we should document, btw https://github.com/kata-containers/documentation/pull/565/files is a PR from @jodh-intel that never landed, you may want to used as base for your document

@devimc yeah that's expected because of the permission mode of kvm node. Is the kvm permission bit distro dependent?
Mine is Ubuntu 18.04.5 LTS and the permission of `/dev/kvm` is:
```
Feng.Wang@ip-10-110-22-245:~$ ls -la /dev/kvm
crw-rw---- 1 root kvm 10, 232 Aug 17 15:20 /dev/kvm
```

devimc commented:
/dev/kvm group owner is set by a udev rule that is added when qemu is installed from distro's packages, since  I didn't installed qemu from distro's packages I don't have that rule

fengwang666 commented:
I have been testing different use cases with this change. I have verified networking and directory based volume mount are working with non-root qemu. However, I'm having difficulty to get device based volume mount to work.

I use a K8s cluster for the testing. 

deployment.yaml:
```
 apiVersion: apps/v1                                                                                                                                                                                                                   
 kind: Deployment                                                                                                                                                                                                                      
 metadata:                                                                                                                                                                                                                             
   name: nginx-deployment                                                                                                                                                                                                              
 spec:                                                                                                                                                                                                                                 
   selector:                                                                                                                                                                                                                           
     matchLabels:                                                                                                                                                                                                                      
       app: nginx                                                                                                                                                                                                                      
   replicas: 2                                                                                                                                                                 
   template:                                                                                                                                                                                                                           
     metadata:                                                                                                                                                                                                                         
       labels:                                                                                                                                                                                                                         
         app: nginx                                                                                                                                                                                                                    
     spec:                                                                                                                                                                                                                             
       runtimeClassName: kata-runtime                                                                                                                                                                                                  
       containers:                                                                                                                                                                                                                     
       - name: busybox                                                                                                                                                                                                                 
         image: busybox                                                                                                                                                                                                                
         volumeMounts:                                                                                                                                                                                                                 
         - mountPath: /dev                                                                                                                                                                                                             
           name: hostdev                                                                                                                                                                                                               
         ports:                                                                                                                                                                                                                        
         - containerPort: 80                                                                                                                                                                                                           
       volumes:                                                                                                                                                                                                                        
       - name: hostdev                                                                                                                                                                                                                 
         hostPath:                                                                                                                                                                                                                     
           path: /dev/nvme5n1                                                                                                                                                                                                          
           type: BlockDevice  
```
Initially I got permission denied error. After I relaxed the permission of `/dev/nvme5n1` to 666, then I got a different error message from kata agent.
```
Aug 20 04:37:57 ip-192-168-0-67 kata[3719]: time="2021-08-20T04:37:57.707881326Z" level=error msg="createContainer failed" error="rpc error: code = Internal desc = Invalid argument (os error 22)" name=containerd-shim-v2 pid=3719 sandbox=7eddaa5790ff34c25bdc5f4b629a4fba137138c5e500fc49e85a5bb6ac902fa8 source=virtcontainers subsystem=kata_agent
```

Even after I switch the code path back to use a root QEMU, the error still exists. Am I testing this correctly? Where can I find the detail log from the kata agent? @bergwolf @devimc @egernst any pointers?

devimc commented:
@fengwang666 

> Even after I switch the code path back to use a root QEMU, the error still exists. Am I testing this correctly? Where can I find the detail log from the kata agent? @bergwolf @devimc @egernst any pointers?

https://github.com/kata-containers/kata-containers/blob/main/docs/Developer-Guide.md#enable-full-debug

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,99 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+
+	if err != nil {
+		return "", err
+	} else {
```
this `else` is not needed, `return userName, nil` should be enough

fengwang666 replied:
Updated.

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,99 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+	if _, err := utils.RunCommand([]string{"/usr/sbin/userdel", "-f", user}); err != nil {
+		shimLog.WithField("username", user).WithError(err).Warn("failed to remove user")
+	}
+}
```
this is weird, do we need a new line ?

fengwang666 replied:
What is weird? @devimc

devimc replied:
No new line at the end of file

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,99 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+}
+
+func removeVmmUser(user string) {
+	if _, err := utils.RunCommand([]string{"/usr/sbin/userdel", "-f", user}); err != nil {
```
maybe we should look for `userdel` (/bin, /sbin, /usr/sbin ?) before using it

devimc reviewed (commented):
thanks @fengwang666

devimc commented on src/runtime/virtcontainers/utils/utils.go:
```diff
@@ -390,3 +391,63 @@ outer:
+	if err != nil {
+		return fmt.Errorf("parent directory does not exist %s", filepath.Dir(path))
+	}
+	if stat, ok := info.Sys().(*syscall.Stat_t); ok {
+		if err = syscall.Chown(path, int(stat.Uid), int(stat.Gid)); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("fail to retrieve the uid and gid of path %s", path)
+	}
+
+	return nil
```
I think you can simplify this chunk of code

```go
if stat, ok := info.Sys().(*syscall.Stat_t); ok {
	if err = syscall.Chown(path, int(stat.Uid), int(stat.Gid)); err != nil {
		return err
	}
	return nil
}

return fmt.Errorf("fail to retrieve the uid and gid of path %s", path)
```

fengwang666 replied:
Updated. Thanks!

devimc commented on src/runtime/virtcontainers/utils/utils.go:
```diff
@@ -390,3 +391,63 @@ outer:
+	var sb strings.Builder
+	sb.WriteString(path)
+	// Append '/' to the end if the path doesn't end with '/'
+	if !os.IsPathSeparator(path[len(path)-1]) {
```
can we change this to:

```go
if !strings.HasSuffix(path, string(os.PathSeparator)) {
```

fengwang666 commented:
@devimc @egernst I have addressed all the comments in the PR. Since additional unit tests would likely require root user to run, I'll add more unit tests once https://github.com/kata-containers/kata-containers/pull/2447 is merged in. PTAL. Thanks!

bergwolf commented:
Thanks @fengwang666 ! Do we have a CI test to validate this? It is prone to breakage without a CI.

bergwolf commented:
/test

fengwang666 commented:
@bergwolf, @devimc has an open [PR](https://github.com/kata-containers/tests/pull/3846) that adds a CI for this feature. Can rerun the workflow? I need to update the vendor to bring in the govmm change.

devimc commented on src/runtime/cli/config/configuration-qemu.toml.in:
```diff
@@ -24,6 +24,11 @@ machine_type = "@MACHINETYPE@"
+# Enable running QEMU VMM as a non-root user.
+# By default QEMU VMM run as root. When this is set to true, QEMU VMM process runs as
+# a non-root random user. See documentation for the limitations of this mode.
+# rootless = true
```
do we need an annotation for this?  this way rootless can be enable or disable through k8s annotations, wdyt?

fengwang666 replied:
I don't have a preference. Maybe we can defer this to a later time?

devimc replied:
ok, just file an issue please

fengwang666 replied:
Updated.

amshinde replied:
As I was reviewing your PR for adding documentation (https://github.com/kata-containers/kata-containers/pull/2546), I am finding this terminology a bit confusing. 
I would like to make sure we clearly make the distinction between true `rootless` mode where the runtime itself is invoked as rootless from the upper stack vs what is being done here i.e running the hypervisor as rootless.

I would like to propose we name this configuration something else to avoid confusion to other users, something like `rootless-hypervisor`. 
Would like to see what others think about this.
cc @jodh-intel @egernst @fidencio @bergwolf @c3d

devimc replied:
all options in this section (`[hypervisor]`)  **only** have effect on the hypervisor, so adding suffixes like `hypervisor` or `vmm`  are useless / pointless.

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,104 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+		return err
+	}
+
+	info, _ := os.Stat("/dev/kvm")
+	var Gid uint32
+	if stat, ok := info.Sys().(*syscall.Stat_t); ok {
+		Gid = stat.Gid
+	} else {
+		return fmt.Errorf("failed to get the gid of /dev/kvm")
+	}
+	// Add the kvm group to the hypervisor supplemental group so that the hypervisor process can access /dev/kvm
+	runtimeConfig.HypervisorConfig.Groups = append(runtimeConfig.HypervisorConfig.Groups, Gid)
+
+	return nil
```
this will crash on systems where kvm modules is not loaded, let's check the error returned by `os.Stat` and simplify the code a little bit more, wdyt?

```go
info, err := os.Stat("/dev/kvm")
if err != nil {
	return err
}

if stat, ok := info.Sys().(*syscall.Stat_t); ok {
	// Add the kvm group to the hypervisor supplemental group so that the hypervisor process can access /dev/kvm
	runtimeConfig.HypervisorConfig.Groups = append(runtimeConfig.HypervisorConfig.Groups, stat.Gid)
	return nil
}
return fmt.Errorf("failed to get the gid of /dev/kvm")
```

devimc commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,104 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+		return "", err
+	}
+
+	maxAttempt := 5
+	attempt := 0
+	for ok := true; ok; ok = err != nil && attempt < maxAttempt {
+		userName = fmt.Sprintf("qemu-%v", rand.Intn(100000))
+		_, err = utils.RunCommand([]string{useraddPath, "-M", "-s", nologinPath, userName})
+		attempt++
+		if err != nil {
+			shimLog.WithField("attempt", attempt).WithField("username", userName).
+				WithError(err).Warn("failed to add user, will try again")
+		}
+	}
+	if err != nil {
+		return "", err
+	}
+	return userName, nil
```
this chuck of code is not easy to read to me, can we simplify it?

```go
	maxAttempt := 5
	for i := 0; i < maxAttempt; i++ {
		userName = fmt.Sprintf("qemu-%v", rand.Intn(100000))
		_, err = utils.RunCommand([]string{useraddPath, "-M", "-s", nologinPath, userName})
		if err == nil {
			return userName, nil
		}
	}
	return "", fmt.Errorf("Could not create VMM user: %v", err)
```

devimc reviewed (commented):
thanks @fengwang666 - I have more comments

devimc commented:
btw CI is failing

fengwang666 commented on src/runtime/virtcontainers/virtiofsd.go:
```diff
@@ -79,6 +79,10 @@ func (v *virtiofsd) getSocketFD() (*os.File, error) {
 		return nil, err
 	}
 
+	if err := utils.ChownToParent(v.socketPath); err != nil {
```
`virtiofsd` is still running as root. We briefly discussed this a few weeks ago in the AC meeting. It would be nice to also run `virtiofsd` as a non-root process, but we agreed that's a separate task and can be triaged separately.

devimc replied:
ok, if that's the case, would you mind filing a bug and commenting it in the code

fengwang666 replied:
https://github.com/kata-containers/kata-containers/issues/2542 created. I'll add a comment in the code as well.

devimc commented:
@fengwang666 rebase please

devimc commented:
To fix the alignment issue: 

```patch
diff --git a/src/runtime/pkg/katautils/config.go b/src/runtime/pkg/katautils/config.go
index edf882c5..5f979e91 100644
--- a/src/runtime/pkg/katautils/config.go
+++ b/src/runtime/pkg/katautils/config.go
@@ -105,12 +105,12 @@ type hypervisor struct {
        EnableAnnotations       []string `toml:"enable_annotations"`
        RxRateLimiterMaxRate    uint64   `toml:"rx_rate_limiter_max_rate"`
        TxRateLimiterMaxRate    uint64   `toml:"tx_rate_limiter_max_rate"`
+       MemOffset               uint64   `toml:"memory_offset"`
        VirtioFSCacheSize       uint32   `toml:"virtio_fs_cache_size"`
        NumVCPUs                int32    `toml:"default_vcpus"`
        DefaultMaxVCPUs         uint32   `toml:"default_maxvcpus"`
        MemorySize              uint32   `toml:"default_memory"`
        MemSlots                uint32   `toml:"memory_slots"`
-       MemOffset               uint64   `toml:"memory_offset"`
        DefaultBridges          uint32   `toml:"default_bridges"`
        Msize9p                 uint32   `toml:"msize_9p"`
        PCIeRootPort            uint32   `toml:"pcie_root_port"`
```

egernst commented:
/test

egernst commented:
/test

egernst commented:
/test

devimc commented on src/runtime/virtcontainers/qemu.go:
```diff
@@ -1007,6 +1014,26 @@ func (q *qemu) cleanupVM() error {
+			q.Logger().WithError(err).WithField("uid", q.config.Uid).Warn("failed to find the user")
+		}
+
+		_, err = pkgUtils.RunCommand([]string{"userdel", "-f", u.Username})
```
please use `utils.FirstValidPath(`

devimc reviewed (commented):
thanks @fengwang666  - I've left one comment

devimc commented:
@fengwang666 I'm trying to run a rootless container in a new VM, where `/run/kata-containers` and `/run/vc` don't exist but I'm getting the following error:

```
FATA[0000] run pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: failed to launch qemu: exit status 1, error messages from qemu log: qemu-system-x86_64: cannot create PID file: Cannot open pid file: Permission denied
: unknown 
[run.sh:37] ERROR: sudo crictl runp -r kata integration/rootless/rootless-pod.json
Makefile:174: recipe for target 'rootless' failed
make: *** [rootless] Error 1

```

any thoughts?

devimc commented:
@fidencio @jodh-intel would you mind reviewing this?  `MkdirAllWithInheritedOwner` function worries me, looks insecure :worried: , ptal

fengwang666 commented:
> @fengwang666 I'm trying to run a rootless container in a new VM, where `/run/kata-containers` and `/run/vc` don't exist but I'm getting the following error:
> 
> ```
> FATA[0000] run pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: failed to launch qemu: exit status 1, error messages from qemu log: qemu-system-x86_64: cannot create PID file: Cannot open pid file: Permission denied
> : unknown 
> [run.sh:37] ERROR: sudo crictl runp -r kata integration/rootless/rootless-pod.json
> Makefile:174: recipe for target 'rootless' failed
> make: *** [rootless] Error 1
> ```
> 
> any thoughts?

I have not seen this before. Does it fail consistently?

devimc commented:
/test

devimc commented:
/test-debian
/test-fedora

jodh-intel commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,98 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+		return "", err
+	}
+
+	maxAttempt := 5
```
Can you add comments explaining why we need to make multiple attempts here? Since the username is random, I'm assuming this is to handle collision where we try to add an already existing user. But there is also the potential scenario that we cannot create the user because another instance of the runtime is already doing it, or an admin is running a command like `vipw`/`vigrp`.

jodh-intel commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,98 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+	if err != nil {
+		shimLog.WithField("username", user).WithError(err).Warn("failed to remove user")
+	}
+	if _, err := utils.RunCommand([]string{userdelPath, "-f", user}); err != nil {
```
No collision possibility here, but possible a locking one so shouldn't there be multiple attempts at removing the user too?

fengwang666 replied:
Considering that this path is only called when it's handling the error and the chance of locking is small, I don't feel strongly we need to add retries here.

fidencio replied:
I second on just having the same multiple attemps logic it.  It doesn't hurt having it. :-)

jodh-intel commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,98 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+
+	maxAttempt := 5
+	for i := 0; i < maxAttempt; i++ {
+		userName = fmt.Sprintf("qemu-%v", rand.Intn(100000))
```
I appreciate this PR is specifically for QEMU, but this code is generic, so maybe `kata-%v` would be better.

jodh-intel commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,98 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+	maxAttempt := 5
+	for i := 0; i < maxAttempt; i++ {
+		userName = fmt.Sprintf("qemu-%v", rand.Intn(100000))
+		_, err = utils.RunCommand([]string{useraddPath, "-M", "-s", nologinPath, userName})
```
Might be worth adding something like, `-c "Kata Containers temporary hypervisor user"`?

jodh-intel commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,98 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+		userName string
+	)
+
+	useraddPath, err := utils.FirstValidPath([]string{"/usr/sbin/useradd", "/sbin/useradd", "/bin/useradd"})
```
I like the simplicity of this, however I can't help feeling a more ephemeral approach using namespaces would be better as they wouldn't persist.

fengwang666 replied:
@jodh-intel Can you elaborate a bit on the namespace approach?

jodh-intel replied:
We'd need to think carefully about this, but I personally don't like the idea of creating unique users due to the potential locking issue and the fact that the users may end up not being deleted (if the runtime is killed for example). That would create an administrative headache so I'd much prefer we constrain the process in a private user namespace, but what do others think?

jodh-intel replied:
To be clear, I personally don't mind us creating a _single_ `kata` user and group (although that will need to be documented carefully), ... just not potentially hundreds or even thousands of them :smile:

jodh-intel commented on src/runtime/pkg/utils/utils.go:
```diff
@@ -80,3 +80,16 @@ func EnsureDir(path string, mode os.FileMode) error {
+func FirstValidPath(paths []string) (string, error) {
+	for _, p := range paths {
+		if _, err := os.Stat(p); err != nil {
+			if os.IsNotExist(err) {
```
Nit: There is a small chance that an admin may have disabled execute permission on one of the (valid) binaries so although it exists, it won't work. That would defeat this function so can you check that the file is also executable?

jodh-intel commented on src/runtime/virtcontainers/utils/utils.go:
```diff
@@ -390,3 +391,61 @@ outer:
+	sb.WriteString(path)
+	// Append '/' to the end if the path doesn't end with '/'
+	if !strings.HasSuffix(path, string(os.PathSeparator)) {
+		sb.WriteRune(os.PathSeparator)
```
I'm a bit confused by this function tbh. Why can't you use `[file]path.Join()` ooi?

fengwang666 replied:
The purpose is to format the path to `/foo/bar/` so that the following code doesn't need to handle the case where a path doesn't end with `/`.

jodh-intel replied:
But `filepath.Join()` handles that scenario automatically doesn't it?

jodh-intel commented on src/runtime/virtcontainers/utils/utils.go:
```diff
@@ -390,3 +391,61 @@ outer:
+func ChownToParent(path string) error {
+	info, err := os.Stat(filepath.Dir(path))
+	if err != nil {
+		return fmt.Errorf("parent directory does not exist %s", filepath.Dir(path))
```
I'd consider adding the original error to the returned error since this message _may_ be misleading in certain scenarios.

jodh-intel commented on src/runtime/virtcontainers/utils/utils.go:
```diff
@@ -390,3 +391,61 @@ outer:
+	return nil
+}
+
+func ChownToParent(path string) error {
```
- Can you document this function. Is path supposed to be an absolute? Are symlinks handled?
- Paths can be awkward and dangerous, so I'd consider checking that `path`:
  - is not `""`.
  - starts with a slash.

Same comments for `MkdirAllWithInheritedOwner()` above.

egernst replied:
@fengwang666  -- I think we still want to address MkdirAllWithInheritedOwner ?

fengwang666 replied:
fixed

jodh-intel commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,98 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+	}
+	defer func() {
+		if err != nil {
+			removeVmmUser(userName)
```
Shouldn't this be called when the shim ends too, to avoid leaving an unusable temporary user in the password file?

fengwang666 replied:
The created user is deleted in the [cleanupVM()](https://github.com/fengwang666/kata-containers/blob/6ddee62128e10cef37259697224593b8cc53865a/src/runtime/virtcontainers/qemu.go#L1030-L1037) call. I didn't reuse the code because the logger is different.

jodh-intel reviewed (commented):
Thanks @fengwang666.

bergwolf commented:
/test

fengwang666 commented:
/test

fidencio commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,100 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+		if err = os.Mkdir(userTmpDir, vc.DirMode); err != nil {
+			return err
+		}
+		if err = syscall.Chown(userTmpDir, uid, gid); err != nil {
```
Aren't we leaving userTmpDir behind in case it fails?  And this applies to every single place where we return err in this function.

egernst replied:
this should be handled by defer.

fidencio commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,100 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+		userName string
+	)
+
+	useraddPath, err := utils.FirstValidExecutable([]string{"/usr/sbin/useradd", "/sbin/useradd", "/bin/useradd"})
```
This worries me a little bit, but we discussed this during the AC meeting and seems folks are fine about it.

fidencio commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,100 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+	if err != nil {
+		return "", err
+	}
+	nologinPath, err := utils.FirstValidExecutable([]string{"/usr/sbin/nologin", "/sbin/nologin", "/bin/nologin"})
```
This also worries me a little bit, but we discussed this during the AC meeting and seems folks are fine about it.

fidencio commented on src/runtime/containerd-shim-v2/create.go:
```diff
@@ -259,3 +272,100 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
+		return fmt.Errorf("%s is expected to be a directory", userTmpDir)
+	}
+
+	if err := os.Setenv("XDG_RUNTIME_DIR", userTmpDir); err != nil {
```
Okay, you overwrite XDG_RUNTIME_DIR here in order to be able to cleanup userTempDir later on.

Still, in case of failures it'd be leaked.

fidencio commented on src/runtime/virtcontainers/qemu.go:
```diff
@@ -1007,6 +1014,29 @@ func (q *qemu) cleanupVM() error {
 
+	if rootless.IsRootless() {
+		rootlessDir := os.Getenv("XDG_RUNTIME_DIR")
+		if rootlessDir != "" {
```
I get that you set XDG_RUNTIME_DIR and here you're basically ensuring that your userTmpDir was created and then removing it.  However, what's the case that `os.Getenv("XDG_RUNTIME_DIR")` would return `""`?

I think this check may be a no-op and we'd endup either deleting the userTmpDir or whater is under `/run/user/uid`.  Am I missing something? (It's very much possible, by the way ;-))

fengwang666 replied:
You're right. The empty string check is not necessary. Removed.

fidencio commented on src/runtime/pkg/utils/utils.go:
```diff
@@ -80,3 +80,21 @@ func EnsureDir(path string, mode os.FileMode) error {
+			return "", err
+		}
+		mode := info.Mode()
+		if mode&0111 == 0 {
```
Please, add a comment here on what you're checking.

fidencio reviewed (commented):
I have added a few comments and I'd really like a second review from @amshinde @devimc @jodh-intel @egernst.

I'm not that much comfortable with this approach, but I see this as something we can expand on as long as we really document this as something "experimental" for now (or, IOW, don't use it unless you really know what you're doing).

@fengwang666, considering it'll get approved, I'd like to see https://github.com/kata-containers/kata-containers/pull/2432/commits/96943be1b78f7bff5c482d05ace9b8d18d7b5a1d properly broken up and "squashed" to the first patch of the series.

Thanks for the contribution, and I'll revisit this Tomorrow.

fidencio commented:
While here, @bergwolf, @sameo, @sbrivio-rh, if you happen to have cycles I'd really like to hear from you.

fengwang666 commented:
/test

fengwang666 commented:
/test

fengwang666 commented:
/test-debian
/test-fedora

fengwang666 commented:
/test-debian

fengwang666 commented:
/test-fc

fengwang666 commented:
/test-debian

sbrivio-rh commented:
Hi,

> While here, @bergwolf, @sameo, @sbrivio-rh, if you happen to have cycles I'd really like to hear from you.

Sorry for the late reply. I have some questions:

* what is the advantage of this approach compared to the somewhat more common way of running qemu as `nobody:kvm`?
* did you consider, as a replacement or as an addition, to use the `-runas` qemu command line parameter (qemu would drop privileges itself)? As an addition, I _guess_ this allows to drop the `kvm` group ID too, because the `/dev/kvm` device node would be opened before the guest starts, and `-runas` takes effect immediately before the guest is started -- I haven't tested whether this actually works, though
* in containers terminology, "rootless" usually implies a completely unprivileged deployment. Isn't calling the configuration flag `rootless` a bit confusing in this sense? That might seem to suggest that Kata Containers support a rootless mode (especially in terms of networking setup)
* shouldn't this be the default way the runtime starts qemu? That is, what is the advantage of running qemu as root by default?

fengwang666 commented:
> > * did you consider, as a replacement or as an addition, to use the `-runas` qemu command line parameter (qemu would drop privileges itself)? As an addition, I _guess_ this allows to drop the `kvm` group ID too, because the `/dev/kvm` device node would be opened before the guest starts, and `-runas` takes effect immediately before the guest is started -- I haven't tested whether this actually works, though
> 
> I just saw Paolo mentioning `systemd-run`, and that may also be an approach to consider.

What benefits would the `systemd-run` provide compared to the current approach?

fidencio commented:
> > > * did you consider, as a replacement or as an addition, to use the `-runas` qemu command line parameter (qemu would drop privileges itself)? As an addition, I _guess_ this allows to drop the `kvm` group ID too, because the `/dev/kvm` device node would be opened before the guest starts, and `-runas` takes effect immediately before the guest is started -- I haven't tested whether this actually works, though
> > 
> > 
> > I just saw Paolo mentioning `systemd-run`, and that may also be an approach to consider.
> 
> What benefits would the `systemd-run` provide compared to the current approach?

I don't know, I didn't do the comparison.  I just threw another idea in the bucket of ideas we have here.  It may or may not be worth exploring `systemd-run` or `-runas`.

fengwang666 commented:
> I don't know, I didn't do the comparison. I just threw another idea in the bucket of ideas we have here. It may or may not be worth exploring `systemd-run` or `-runas`.

@fidencio IIUC, both approaches (`systemd-run` or `-runas`) would still require Kata to create the temporary user and group, which seems to be the main concern of the current approach.

fengwang666 commented:
> Hi,
> 
> > While here, @bergwolf, @sameo, @sbrivio-rh, if you happen to have cycles I'd really like to hear from you.
> 
> Sorry for the late reply. I have some questions:
> 
> * what is the advantage of this approach compared to the somewhat more common way of running qemu as `nobody:kvm`?

In our use cases, there are multiple Kata Pods running on the same machine. And we prefer to have each hypervisor process running as a different user so that they cannot access each other's files in case a VM breakout happens.

> * did you consider, as a replacement or as an addition, to use the `-runas` qemu command line parameter (qemu would drop privileges itself)? As an addition, I _guess_ this allows to drop the `kvm` group ID too, because the `/dev/kvm` device node would be opened before the guest starts, and `-runas` takes effect immediately before the guest is started -- I haven't tested whether this actually works, though

This might work. It would require some additional testing.

> * in containers terminology, "rootless" usually implies a completely unprivileged deployment. Isn't calling the configuration flag `rootless` a bit confusing in this sense? That might seem to suggest that Kata Containers support a rootless mode (especially in terms of networking setup)

The flag `rootless` is the under the hypervisor section of the configuration. I agree it can cause confusion though. Any other suggestion on the name?

> * shouldn't this be the default way the runtime starts qemu? That is, what is the advantage of running qemu as root by default?

Certain features (e.g. attach block devices) as described in the [doc](https://github.com/kata-containers/kata-containers/pull/2546) are not supported in the current non-root QEMU mode. It might break some existing customers if we switch this to default. 

Thanks. @sbrivio-rh

devimc commented:
/test-vfio
/test-ubuntu-metrics

fengwang666 commented:
/test

fengwang666 commented:
/test

fengwang666 commented:
> Hi,
> 
> > While here, @bergwolf, @sameo, @sbrivio-rh, if you happen to have cycles I'd really like to hear from you.
> 
> Sorry for the late reply. I have some questions:
> 
> * what is the advantage of this approach compared to the somewhat more common way of running qemu as `nobody:kvm`?
> * did you consider, as a replacement or as an addition, to use the `-runas` qemu command line parameter (qemu would drop privileges itself)? As an addition, I _guess_ this allows to drop the `kvm` group ID too, because the `/dev/kvm` device node would be opened before the guest starts, and `-runas` takes effect immediately before the guest is started -- I haven't tested whether this actually works, though
> * in containers terminology, "rootless" usually implies a completely unprivileged deployment. Isn't calling the configuration flag `rootless` a bit confusing in this sense? That might seem to suggest that Kata Containers support a rootless mode (especially in terms of networking setup)
> * shouldn't this be the default way the runtime starts qemu? That is, what is the advantage of running qemu as root by default?

@sbrivio-rh
I tried launching QEMU with the `-runas` flag (the [commit](https://github.com/kata-containers/kata-containers/pull/2432/commits/ed300fba0cb5ac153dac13b63185309fd8408152)), but the QEMU fails to start with segfault:
```
Sep 14 20:26:44 ip-10-110-22-245 kernel: [95831.828294] qemu-system-x86[86375]: segfault at 10 ip 00007ff4f1f6cd8d sp 00007ff4f16a0a20 error 4
Sep 14 20:26:44 ip-10-110-22-245 kernel: [95831.828299] qemu-system-x86[86363]: segfault at 10 ip 00007ff4f1f6cd8d sp 00007ff4f3e37ae0 error 4
Sep 14 20:26:44 ip-10-110-22-245 kernel: [95831.828301]  in libpthread-2.27.so[7ff4f1f67000+1a000]
Sep 14 20:26:44 ip-10-110-22-245 kernel: [95831.828308] Code: 89 f3 48 83 ec 08 8b 6e 10 e8 4f fa ff ff 39 c5 74 0b 48 83 c4 08 5b 5d c3 0f 1f 40 00 83 7b 08 fa 75 ef 48 8b 05 43 86 21 00 <48>
 8b 70 10 48 8b 78 08 48 8b 50 18 8b 00 0f 05 48 8b 3d 2c 86 21
Sep 14 20:26:44 ip-10-110-22-245 kernel: [95831.828309]  in libpthread-2.27.so[7ff4f1f67000+1a000]
Sep 14 20:26:44 ip-10-110-22-245 kernel: [95831.828315] Code: 89 f3 48 83 ec 08 8b 6e 10 e8 4f fa ff ff 39 c5 74 0b 48 83 c4 08 5b 5d c3 0f 1f 40 00 83 7b 08 fa 75 ef 48 8b 05 43 86 21 00 <48>
 8b 70 10 48 8b 78 08 48 8b 50 18 8b 00 0f 05 48 8b 3d 2c 86 21
Sep 14 20:26:44 ip-10-110-22-245 kernel: [95831.828319] qemu-system-x86[86373]: segfault at 10 ip 00007ff4f1f6cd8d sp 00007ff4f1ea1b20 error 4 in libpthread-2.27.so[7ff4f1f67000+1a000]
Sep 14 20:26:44 ip-10-110-22-245 kernel: [95831.828328] Code: 89 f3 48 83 ec 08 8b 6e 10 e8 4f fa ff ff 39 c5 74 0b 48 83 c4 08 5b 5d c3 0f 1f 40 00 83 7b 08 fa 75 ef 48 8b 05 43 86 21 00 <48>
 8b 70 10 48 8b 78 08 48 8b 50 18 8b 00 0f 05 48 8b 3d 2c 86 21
```

fengwang666 commented:
/test

fengwang666 commented:
/test

fengwang666 commented:
/retest-debian
/retest-fedora
/retest-centos

egernst commented on src/runtime/virtcontainers/hypervisor.go:
```diff
@@ -221,6 +221,7 @@ type Param struct {
 }
 
 // HypervisorConfig is the hypervisor configuration.
+// nolint: govet
```
Any reason to not have this fixed now? I worry about carrying this debt.

fidencio replied:
Same here.

fengwang666 commented:
/test

fengwang666 commented:
/rest

fengwang666 commented:
/retest

fengwang666 commented:
/retest-fedora

fengwang666 commented:
/retest-debian
/retest-centos

sbrivio-rh commented:
> > Hi,
> > > While here, @bergwolf, @sameo, @sbrivio-rh, if you happen to have cycles I'd really like to hear from you.
> > 
> > 
> > Sorry for the late reply. I have some questions:
> > 
> > * what is the advantage of this approach compared to the somewhat more common way of running qemu as `nobody:kvm`?
> 
> In our use cases, there are multiple Kata Pods running on the same machine. And we prefer to have each hypervisor process running as a different user so that they cannot access each other's files in case a VM breakout happens.

But wait, other than the files otherwise introduced with this change, is there anything else that user would write? By the way, that user could be `nobody`.

> > * did you consider, as a replacement or as an addition, to use the `-runas` qemu command line parameter (qemu would drop privileges itself)? As an addition, I _guess_ this allows to drop the `kvm` group ID too, because the `/dev/kvm` device node would be opened before the guest starts, and `-runas` takes effect immediately before the guest is started -- I haven't tested whether this actually works, though
> 
> This might work. It would require some additional testing.
> 
> > * in containers terminology, "rootless" usually implies a completely unprivileged deployment. Isn't calling the configuration flag `rootless` a bit confusing in this sense? That might seem to suggest that Kata Containers support a rootless mode (especially in terms of networking setup)
> 
> The flag `rootless` is the under the hypervisor section of the configuration. I agree it can cause confusion though. Any other suggestion on the name?

Perhaps `drop_root`, but I hope there are better suggestions ;) If it's too ugly, I see the point of it being under the hypervisor section, so perhaps it's not too bad either.

> > * shouldn't this be the default way the runtime starts qemu? That is, what is the advantage of running qemu as root by default?
> 
> Certain features (e.g. attach block devices) as described in the [doc](https://github.com/kata-containers/kata-containers/pull/2546) are not supported in the current non-root QEMU mode. It might break some existing customers if we switch this to default.

Oh, I see now, sure, it makes sense.

> I tried launching QEMU with the `-runas` flag (the [commit](https://github.com/fengwang666/kata-containers/commit/ed300fba0cb5ac153dac13b63185309fd8408152)), but the QEMU fails to start with segfault:

I couldn't check with your commit on Kata yet. I just quickly tried this:

```
# id
uid=0(root) gid=0(root) groups=0(root)
# getent group kvm
kvm:x:106:
# id nobody
uid=65534(nobody) gid=65534(nogroup) groups=65534(nogroup)
# kvm -runas 65534:106 -kernel vmlinuz -initrd initrd.img -nographic -nodefaults -serial stdio -append "console=ttyS0"
```
and it works for me. Now, of course it shouldn't crash :) ...does that happen before the guest starts? Does qemu print anything to stdout/stderr before that? I mean, if it doesn't work, never mind -- I just thought it would simplify things quite a bit.

Jakob-Naucke commented:
I can get this to work on k8s, but plain `ctr` fails for me (x86_64 and s390x): `ctr: failed to create shim: the path is not absolute: unknown`. Has anyone else observed this?

fengwang666 commented:
> I can get this to work on k8s, but plain `ctr` fails for me (x86_64 and s390x): `ctr: failed to create shim: the path is not absolute: unknown`. Has anyone else observed this?

@Jakob-Naucke Thanks for reporting this. It's related to the latest change I made to the PR. I'll address it.

fengwang666 commented:
> > I can get this to work on k8s, but plain `ctr` fails for me (x86_64 and s390x): `ctr: failed to create shim: the path is not absolute: unknown`. Has anyone else observed this?
> 
> @Jakob-Naucke Thanks for reporting this. It's related to the latest change I made to the PR. I'll address it.

@Jakob-Naucke I pushed a fix. Feel free to pull the latest and try again. Let me know.

fengwang666 commented:
/test

fengwang666 commented:
/test

Jakob-Naucke commented:
> > > I can get this to work on k8s, but plain `ctr` fails for me (x86_64 and s390x): `ctr: failed to create shim: the path is not absolute: unknown`. Has anyone else observed this?
>
> ...
> @Jakob-Naucke I pushed a fix. Feel free to pull the latest and try again. Let me know.

@fengwang666 awesome, works now

fengwang666 commented:
/test

fengwang666 commented:
/test-ubuntu

fengwang666 commented:
/test-arm

fengwang666 commented:
/retest-arm

fengwang666 commented:
/retest-ubuntu

fengwang666 commented:
/retest-ubuntu

fengwang666 commented:
/retest-arm

fengwang666 commented:
/test-ubuntu

fengwang666 commented:
/test-arm

fengwang666 commented:
/test-ubuntu
/test-arm

fengwang666 commented:
/test

chavafg commented:
/test-metrics-ubuntu
/test-vfio

chavafg commented:
/test-ubuntu-metrics

fengwang666 commented:
@egernst @fidencio All the CIs have passed. Can you approve and merge it?

fidencio reviewed (approved):
all the comments have been addressed, and we've an agreement on how to add a CI for this in the future.

lgtm, thanks @fengwang666, and let's get this one in for the 2.3.0-alpha1 release.
